First pull request for net-phy-realtek-next

ericwoud opened the pull request:
Take a look, it is not enough tested, don't accept yet.

I have changed the code rtl8221b_vb_cg_match_phy_device() a bit to store the id.

The ID is used for matching in phylink_phy_no_inband()

Also reduce the times phy_read_mmd() / phy_read() is being used, i assumed the id on the early 8221b itself equals zero, so only read it when the id == 0. I do not have such a rtl8221b so I do not know if this works.

The rest of the code is pretty straight forward.

I think `genphy_c45_pma_resume(phydev)` can be removed from rtl822x_c45_config_init(), as it will be called later anyway, but I need to test that.

dangowrt committed:
net: phy: realtek: detect early version of RTL8221B

Early versions (?) of the RTL8221B PHY cannot be identified in a regular
Clause-45 bus scan as the PHY doesn't report the implemented MMDs
correctly but returns 0 instead.
Implement custom identify function using the PKGID instead of iterating
over the implemented MMDs.

Signed-off-by: Daniel Golle <daniel@makrotopia.org>

ericwoud committed:
net: phy: sfp: Fixup for OEM SFP-2.5G-T module

Implement fixup for OEM SFP-2.5G-T module. It can now recognise and attach
the rtl8221B, using the RollBall protocol.

Remove the quirk. With implementing the fixup, the PHY is attached, so the
values changed in the quirk are no longer used.

ericwoud committed:
net: phy: realtek: rtlgen_get_speed(): Add 2500-Lite mode

Add 2500-Litemode, reporting 1000 Mbps.

ericwoud committed:
net: phy: phy_device: Skip probe/read id from Realtek MDIO_MMD_VEND1

Realtek PHY do not support reading from MDIO_MMD_VEND1 at MDIO_STAT2
or MII_PHYSIDx. The registers are undocumented in the datasheet and
reading from them makes the PHY behave erratically and fail.
Only garbage is read from the ID register, a random number.

ericwoud committed:
net: phy: realtek: rtlgen_get_speed(): Pass register value as argument

The register that needs to be read to determine the speed, is not
always read in the same manner, also not always at the same address.

To be able to use the one rtlgen_get_speed() function, generalize
rtlgen_get_speed() even more. Pass the value from the device specific
register as argument to rtlgen_get_speed().

dangowrt commented on drivers/net/phy/realtek.c:
```diff
@@ -1249,6 +1406,16 @@ static struct phy_driver realtek_drvs[] = {
+		.get_features   = rtl822x_c45_get_features,
+		.config_aneg    = rtl822x_c45_config_aneg,
+		.read_status    = rtl822x_c45_read_status,
+		.suspend        = rtl822x_c45_suspend,
```
Why not assign `genphy_c45_pma_suspend` directly here?

ericwoud replied:
I did have some code here, and used it for debugging messages, but now indeed it is empty and I planned to remove , if it would stay like this. So agree.

dangowrt commented on drivers/net/phy/realtek.c:
```diff
@@ -1249,6 +1406,16 @@ static struct phy_driver realtek_drvs[] = {
+		.config_aneg    = rtl822x_c45_config_aneg,
+		.read_status    = rtl822x_c45_read_status,
+		.suspend        = rtl822x_c45_suspend,
+		.resume         = rtl822x_c45_resume,
```
Same here, we could get rid of `rtl822x_c45_resume` by just setting `genphy_c45_pma_suspend`

ericwoud replied:
I did have some code here, and used it for debugging messages, but now indeed it is empty and I planned to remove , if it would stay like this. So agree, I only need to use it once to check the order of execution .config_init and ,resume, to see if I need to call genphy_c45_pma_resume() from rtl822x_c45_config_init().

dangowrt commented on drivers/net/phy/realtek.c:
```diff
@@ -819,7 +951,36 @@ static int rtl8221b_vb_cg_match_phy_device(struct phy_device *phydev)
+*/
+static bool rtl822x_use_c45(struct phy_device *phydev)
+{
+        return !phydev->mdio.bus->read && !!phydev->mdio.bus->read_c45;
```
Let's just use `phydev->is_c45` here like you are suggesting in the comment

ericwoud replied:
Ok, then can remove the whole `static bool rtl822x_use_c45(struct phy_device *phydev) { }`

ericwoud replied:
Edit: 

Note that in the C45 driver instance, I have not defined `.soft_reset` as the PHY is setup already. There is no C45 version of genphy_soft_reset()

I believe we need to take a close look at the following that has to do with this change as well:

I think that what I fixed in the commit:
```
net: phy: phy_device: Skip probe/read id from Realtek MDIO_MMD_VEND1
```
You fixed the same problem with:
```
net: phy: realtek: use genphy_soft_reset for 2.5G PHYs
```
Where my commit prevents the problem, and your commits clears the problem after it occurred.

But, if you change the early 8221b to use the C45 driver instance, then both fixes fall away. The ID is not yet found so MDIO_MMD_VEND1 is read and there is no reset performed.

The question here, should we rely on a soft-reset to clear a problem created by reading from the PHY in an undocumented/unsupport fashion?

Perhapse finding ID by MDIO_MMD_PMAPMD, MDIO_PKGID1/2 can move to the phy_device.c get_phy_c45_ids() function also?

I'd be happy to write something, but I do not have such device, so do not know where get_phy_c45_ids() fails to find the ID. Does it not find any `devs_in_pkg` ? Or does it not read any ID correctly?

Possibly could implement a soft-reset for C45, like so:
```
static int rtl822x_c45_soft_reset(struct phy_device *phydev)
{
	int val, err;

	err = phy_modify_mmd(phydev, MDIO_MMD_AN , MII_BMCR,
			     BMCR_RESET, BMCR_RESET);
	if (err < 0)
		return err;

	err = phy_modify_mmd(phydev, MDIO_MMD_PCS , MII_BMCR,
			     BMCR_RESET, BMCR_RESET);
	if (err < 0)
		return err;

	err = phy_modify_mmd(phydev, MDIO_MMD_PMAPMD , MII_BMCR,
			     BMCR_RESET, BMCR_RESET);
	if (err < 0)
		return err;

	err = phy_read_mmd_poll_timeout(phydev, MDIO_MMD_PMAPMD, MII_BMCR,
			val, !(val & BMCR_RESET), 5000, 100000, true);
	if (err < 0)
		return err;

	err = phy_read_mmd_poll_timeout(phydev, MDIO_MMD_PCS, MII_BMCR,
			val, !(val & BMCR_RESET), 5000, 100000, false);
	if (err < 0)
		return err;

	err = phy_read_mmd_poll_timeout(phydev, MDIO_MMD_AN, MII_BMCR,
			val, !(val & BMCR_RESET), 5000, 100000, false);
	if (err < 0)
		return err;

	return 0;
}
```
But what would be the most correct order?

ericwoud commented:
Do you agree on commit messages?

Do I leave the part below as is? Or do you want to add this part to your patch? Does this work as I expected?

Edit: I have split the commit, so you could leave this part in a separate commit, or squash it with your commit.

```
@@ -795,6 +796,14 @@ static int rtl8221b_vb_cg_match_phy_device(struct phy_device *phydev)
 	int val;
 	u32 id;
 
+	if (phydev->is_c45) {
+		if (phydev->c45_ids.device_ids[1])
+			return phydev->c45_ids.device_ids[1] == RTL_8221B_VB_CG;
+	} else {
+		if (phydev->phy_id)
+			return phydev->phy_id == RTL_8221B_VB_CG;
+	}
+
 	if (phydev->mdio.bus->read_c45) {
 		val = phy_read_mmd(phydev, MDIO_MMD_PMAPMD, MDIO_PKGID1);
 		if (val < 0)
@@ -819,7 +828,14 @@ static int rtl8221b_vb_cg_match_phy_device(struct phy_device *phydev)
 		id |= val;
 	}
 
-	return (id == 0x001cc849);
+	if (id != RTL_8221B_VB_CG) return 0;
+
+	if (phydev->is_c45)
+		phydev->c45_ids.device_ids[1] = id;
+	else
+		phydev->phy_id = id;
+
+	return true;
 }
 
 static int rtl822x_probe(struct phy_device *phydev)
```

ericwoud commented:
> I think `genphy_c45_pma_resume(phydev)` can be removed from rtl822x_c45_config_init(), as it will be called later anyway, but I need to test that.

.config_init() ( phy_init_hw() ) is called from different places in kernel, not always followed by .resume(), so I'll leave it as it is, __not__ removing. `genphy_c45_pma_resume(phydev)` from rtl822x_c45_config_init()

ericwoud committed:
net: phylink: rtl822x no inband negotiation

The rtl8221b does not work with inband negotiation enabled. It
is not enabled on 2500baseX and the driver also disables it on sgmii.

Setup sfp phylink to not use inband negotiation for the rtl8221b.

dangowrt committed:
net: phy: realtek: setup ALDPS on RTL8221b

Setup Link Down Power Saving Mode according the DTS property
just like for RTL821x 1GE PHYs.

Always enable it on a SFP module.

ericwoud commented:
Fixing the ALDPS sadly the pull-request has conflicts.

Fixed using RTL8221B_MMD_PHY_CTRL instead and moving to rtl8221b_config_init(().

Merged the 2 commits matching and using id for early 8221b.

I think if you remove the last 2 commits from your branch, it can merge again.

ericwoud committed:
net: phy: realtek: Add driver instance for RTL8221-VB-CG via C45

Add driver instance for clause-45 communication with RTL8221-VB-CG.
The driver sets up a separate set of function pointers for clause-45
only accessable PHY's.

This new driver instance will be used by the OEM SFP-2.5G-T module. The
RTL8221-VB-CG registers are setup by the onboard microprocessor that is also
responsible for the RollBall protocol. This protocol can use C45
communication only.

ericwoud commented:
@dangowrt Do you think it is ready to submit to netdev/net-next ?

dangowrt commented:
> @dangowrt Do you think it is ready to submit to netdev/net-next ?

Yes, at least we can start to collect comments and then may re-submit the easy patches first and work on what ever they tell us to work on.

And we could pick up this one as well:
https://github.com/openwrt/openwrt/pull/13414/commits/f088f39856b2c0b991f42a9d62ddc822a72de297#diff-9b534b62603210d1dfffb409d05d5fde5302179f24eb83a2c6307a5aabdd1fef

ericwoud commented:
Ok, so do you send them in one patch-set, or do you want to keep it separate?

It was on my todo-list to check if we can do the interrupt handling on the SFP module as well...
